crates/metadata-store/src/omnipaxos/store.rs

```diff
@@ -0,0 +1,1232 @@
+                    self.members.clone(),
+                );
+                let _ = response_tx.send(Ok(response));
+            } else {
+                // latest reconfiguration didn't include this node, fail it so that caller can retry
+                let _ = response_tx.send(Err(JoinClusterError::Internal(format!(
+                    "failed to include node '{}' in new configuration",
+                    member_id
+                ))));
+            }
+        }
+    }
+
+    fn fail_join_callbacks(&mut self, cause: impl Fn() -> JoinClusterError) {
+        for (_, response_tx) in self.pending_join_requests.drain() {
+            let _ = response_tx.send(Err(cause()));
+        }
+    }
+
+    /// Checks whether the given node id with its storage id is part of the current set of members
+    fn is_member(member_id: MemberId, members: &HashMap<NodeId, StorageId>) -> bool {
+        members
+            .get(&member_id.node_id)
+            .is_some_and(|storage_id| *storage_id == member_id.storage_id)
+    }
+}
+
+enum DecidedEntriesResult {
+    Continue,
+    // Become passive
+    Stop(RocksDbStorage<Request>),
+}
+
+#[allow(dead_code)]
+struct Passive {
+    connection_manager: Arc<ArcSwapOption<ConnectionManager<OmniPaxosMessage>>>,
+    rocksdb_storage: RocksDbStorage<Request>,
+    request_rx: RequestReceiver,
+    join_cluster_rx: JoinClusterReceiver,
+    metadata_writer: Option<MetadataWriter>,
+    storage_id: StorageId,
+}
+
+impl Passive {
+    fn new(
+        rocksdb_storage: RocksDbStorage<Request>,
+        connection_manager: Arc<ArcSwapOption<ConnectionManager<OmniPaxosMessage>>>,
+        request_rx: RequestReceiver,
+        join_cluster_rx: JoinClusterReceiver,
+        metadata_writer: Option<MetadataWriter>,
+        storage_id: StorageId,
+    ) -> Self {
+        connection_manager.store(None);
+
+        Passive {
+            connection_manager,
+            rocksdb_storage,
+            request_rx,
+            join_cluster_rx,
+            metadata_writer,
+            storage_id,
+        }
+    }
+
+    async fn run(self) -> anyhow::Result<Active> {
+        debug!("Run as passive metadata store node.");
+
+        let Passive {
+            connection_manager,
+            mut rocksdb_storage,
+            mut request_rx,
+            mut join_cluster_rx,
+            metadata_writer,
+            storage_id,
+        } = self;
+
+        // todo make configurable
+        let mut join_retry_policy = RetryPolicy::exponential(
+            Duration::from_millis(100),
+            2.0,
+            None,
+            Some(Duration::from_secs(5)),
+        )
+        .into_iter();
+
+        rocksdb_storage
+            .set_nodes_configuration(&Metadata::with_current(|m| m.nodes_config_ref()))
+            .map_err(StorageError::from)?;
+        // todo only try joining if MetadataStoreState::Candidate
+        let mut join_cluster = std::pin::pin!(Self::join_cluster(None, storage_id));
```

Comment:
A potential follow-up is to use the `MetadataStoreState` stored in the `NodesConfiguration` to control which nodes should join a metadata store cluster or not. That could give us a declarative way to define the nodes that should run an active metadata store if they succeed at joining the cluster.

Reply:
Relying on the `NodesConfiguration` to store metadata store relevant information (e.g. the `MetadataStoreState` and addresses) makes things considerably harder because of the circular dependency (we need the metadata store to write/read the `NodesConfiguration` but it also controls the behavior of the metadata store). This becomes especially problematic in the case where we fail during a reconfiguration w/o having received the log yet. Especially, if this node is now needed for the metadata store to become available there is no easy way to obtain the `NodesConfiguration` by following the normal join flow.